File: libs/fleet/sdk-bindings/go/examples/app_controlled.go
package main

import (
	"bytes"
	"encoding/json"
	"errors"
	"fmt"
	"io"
	"os"
	"time"

	cyclopssdk "github.com/trycua/cloud/cyclops-cs/sdk-bindings/go"
)

func runAgent(sdk *cyclopssdk.SDK, sandbox cyclopssdk.Sandbox) (any, error) {
	mcp, err := sdk.ServiceClient(sandbox, "mcp")
	if err != nil {
		return nil, err
	}
	body := []byte(`{"jsonrpc":"2.0","id":1,"method":"initialize","params":{"protocolVersion":"2025-03-26","capabilities":{},"clientInfo":{"name":"cyclops-sdk-go-example","version":"0.1.0"}}}`)
	deadline := time.Now().Add(5 * time.Minute)
	for {
		request, err := mcp.NewRequest("POST", "/mcp", bytes.NewReader(body))
		if err != nil {
			return nil, err
		}
		request.Header.Set("accept", "application/json, text/event-stream")
		request.Header.Set("content-type", "application/json")
		response, err := mcp.Do(request)
		if err != nil {
			return nil, err
		}
		responseBody, readErr := io.ReadAll(response.Body)
		response.Body.Close()
		if readErr != nil {
			return nil, readErr
		}
		if response.StatusCode >= 200 && response.StatusCode < 300 {
			return map[string]any{"namespace": sandbox.Namespace, "claim": sandbox.Claim, "sandbox": sandbox.Sandbox, "mcp_status": response.StatusCode}, nil
		}
		transient := response.StatusCode == 502 || response.StatusCode == 503 || response.StatusCode == 504
		if !transient || time.Now().After(deadline) {
			return nil, fmt.Errorf("MCP initialize failed with HTTP %d: %s", response.StatusCode, responseBody)
		}
		time.Sleep(5 * time.Second)
	}
}

func main() {
	namespace := env("CYCLOPS_NAMESPACE", fmt.Sprintf("sdk-example-%08x", uint32(time.Now().UnixNano())))
	sdk, err := cyclopssdk.Connect(cyclopssdk.Configuration{
		BaseURL: env("CUA_BASE_URL", "https://run.cua.ai"),
		OAuth: cyclopssdk.OAuthConfiguration{
			TokenURL: env("CUA_TOKEN_URL", "https://auth.cua.ai/realms/cyclops-cs/protocol/openid-connect/token"),
			ClientID: required("CUA_CLIENT_ID"), ClientSecret: required("CUA_CLIENT_SECRET"),
		},
	})
	must(err)
	defer func() { must(sdk.Close()) }()
	cpuCores := uint32(4)
	memory := "4Gi"
	imagePullSecret := required("CUA_IMAGE_PULL_SECRET")
	protocol := cyclopssdk.PoolSpecServiceProtocolTCP
	pool, err := sdk.CreatePool(cyclopssdk.CreatePoolRequest{
		Namespace: namespace,
		Spec: cyclopssdk.PoolSpec{
			Replicas: 1,
			Services: []cyclopssdk.PoolSpecService{{
				Name: "mcp", TargetPort: 3000, Protocol: &protocol,
			}},
			Template: cyclopssdk.PoolTemplate{
				ContainerDiskImage: required("CUA_IMAGE"), ImagePullSecret: &imagePullSecret, CpuCores: &cpuCores, Memory: &memory,
			},
		},
	})
	must(err)
	defer func() { must(sdk.DeletePool(pool)) }()
	claim, err := sdk.CreateClaim(cyclopssdk.CreateClaimRequest{Pool: pool})
	must(err)
	defer func() { must(sdk.DeleteClaim(claim)) }()
	sandbox, err := sdk.WaitClaim(claim)
	must(err)
	result, err := runAgent(sdk, sandbox)
	must(err)
	output, err := json.Marshal(result)
	must(err)
	fmt.Println(string(output))
}

func env(name, fallback string) string {
	if value := os.Getenv(name); value != "" {
		return value
	}
	return fallback
}
func required(name string) string {
	if value := os.Getenv(name); value != "" {
		return value
	}
	panic(name + " is required")
}
func must(err error) {
	if err != nil && !errors.Is(err, os.ErrProcessDone) {
		panic(err)
	}
}

File: libs/fleet/sdk-bindings/examples/kotlin/AppControlled.kt
import ai.cua.cyclops.sdk.*
import ai.cua.cyclops.sdk.schema.*
import kotlinx.coroutines.runBlocking
import java.util.Collections

private data class Expected(val method: String, val url: String, val headers: List<Pair<String, String>>, val body: ByteArray?, val status: UShort, val response: String)
private class ScriptedHttpClient : HttpClient {
    private val lock = Any()
    private val expected = ArrayDeque(listOf(
        Expected("POST", "https://keycloak.invalid/token", listOf("accept" to "application/json", "content-type" to "application/x-www-form-urlencoded"), "grant_type=client_credentials&client_id=client-id&client_secret=client-secret".encodeToByteArray(), 200u, "{\"access_token\":\"offline-token\",\"expires_in\":3600}"),
        Expected("POST", "https://cyclops.invalid/api/namespaces", listOf("accept" to "application/json", "content-type" to "application/json", "authorization" to "Bearer offline-token"), "{\"name\":\"default\"}".encodeToByteArray(), 201u, "{}"),
        Expected("POST", "https://cyclops.invalid/api/k8s/apis/cua.ai/v1/namespaces/default/osgymworkspacepools", listOf("accept" to "application/json", "content-type" to "application/json", "authorization" to "Bearer offline-token"), "{\"apiVersion\":\"cua.ai/v1\",\"kind\":\"OSGymWorkspacePool\",\"metadata\":{\"namespace\":\"default\",\"name\":\"default\",\"labels\":null},\"spec\":{\"replicas\":1,\"template\":{\"containerDiskImage\":\"registry.example/desktop:offline\"},\"services\":[{\"name\":\"mcp\",\"targetPort\":8080}]},\"status\":null}".encodeToByteArray(), 201u, "{\"apiVersion\":\"cua.ai/v1\",\"kind\":\"OSGymWorkspacePool\",\"metadata\":{\"namespace\":\"default\",\"name\":\"default\",\"labels\":null},\"spec\":{\"replicas\":1,\"template\":{\"containerDiskImage\":\"registry.example/desktop:offline\"},\"services\":[{\"name\":\"mcp\",\"targetPort\":8080}]},\"status\":null}"),
        Expected("POST", "https://cyclops.invalid/api/k8s/apis/osgym.cua.ai/v1alpha1/namespaces/default/osgymsandboxclaims", listOf("accept" to "application/json", "content-type" to "application/json", "authorization" to "Bearer offline-token"), "{\"apiVersion\":\"osgym.cua.ai/v1alpha1\",\"kind\":\"OSGymSandboxClaim\",\"metadata\":{\"namespace\":\"default\",\"name\":\"claim-1\",\"labels\":null},\"spec\":{\"sandboxTemplateRef\":{\"name\":\"default\"}},\"status\":null}".encodeToByteArray(), 201u, "{\"apiVersion\":\"osgym.cua.ai/v1alpha1\",\"kind\":\"OSGymSandboxClaim\",\"metadata\":{\"namespace\":\"default\",\"name\":\"default\",\"labels\":null},\"spec\":{\"sandboxTemplateRef\":{\"name\":\"default\"}},\"status\":{\"phase\":\"Bound\",\"sandbox\":{\"name\":\"offline-sandbox\"}}}"),
        Expected("GET", "https://cyclops.invalid/api/k8s/apis/osgym.cua.ai/v1alpha1/namespaces/default/osgymsandboxclaims/default", listOf("accept" to "application/json", "content-type" to "application/json", "authorization" to "Bearer offline-token"), null, 200u, "{\"apiVersion\":\"osgym.cua.ai/v1alpha1\",\"kind\":\"OSGymSandboxClaim\",\"metadata\":{\"namespace\":\"default\",\"name\":\"default\",\"labels\":null},\"spec\":{\"sandboxTemplateRef\":{\"name\":\"default\"}},\"status\":{\"phase\":\"Bound\",\"sandbox\":{\"name\":\"offline-sandbox\"}}}"),
        Expected("GET", "https://cyclops.invalid/api/k8s/apis/cua.ai/v1/namespaces/default/osgymworkspacepools/default", listOf("accept" to "application/json", "content-type" to "application/json", "authorization" to "Bearer offline-token"), null, 200u, "{\"apiVersion\":\"cua.ai/v1\",\"kind\":\"OSGymWorkspacePool\",\"metadata\":{\"namespace\":\"default\",\"name\":\"default\",\"labels\":null},\"spec\":{\"replicas\":1,\"template\":{\"containerDiskImage\":\"registry.example/desktop:offline\"},\"services\":[{\"name\":\"mcp\",\"targetPort\":8080}]},\"status\":null}"),
        Expected("POST", "https://cyclops.invalid/api/svc/default/offline-sandbox-mcp/mcp", listOf("authorization" to "Bearer offline-token"), "{\"offline\":true}".encodeToByteArray(), 202u, "offline service accepted"),
        Expected("DELETE", "https://cyclops.invalid/api/k8s/apis/osgym.cua.ai/v1alpha1/namespaces/default/osgymsandboxclaims/default", listOf("accept" to "application/json", "content-type" to "application/json", "authorization" to "Bearer offline-token"), null, 204u, ""),
        Expected("DELETE", "https://cyclops.invalid/api/k8s/apis/cua.ai/v1/namespaces/default/osgymworkspacepools/default", listOf("accept" to "application/json", "content-type" to "application/json", "authorization" to "Bearer offline-token"), null, 204u, ""),
        Expected("DELETE", "https://cyclops.invalid/api/namespaces/default", listOf("accept" to "application/json", "content-type" to "application/json", "authorization" to "Bearer offline-token"), null, 204u, "")
    ))
    override suspend fun execute(request: HttpRequest): HttpResponse = synchronized(lock) {
        val item = expected.removeFirst()
        check(request.method == item.method && request.url == item.url)
        check(request.headers.map { it.name to it.value } == item.headers)
        check((request.body == null && item.body == null) || (request.body != null && item.body != null && request.body.contentEquals(item.body)))
        HttpResponse(item.status, emptyList(), item.response.encodeToByteArray())
    }
    fun assertExhausted() = check(expected.isEmpty())
}

fun main() = runBlocking {
    val vmTemplate = VmTemplate("registry.example/desktop:offline", null, null, null, null, null, null, null, null, null, null, null, null, null)
    val spec = PoolSpec(1u, PoolTemplate(null, null, null, null, null, vmTemplate.containerDiskImage, null, null, null, null, null, null), null, listOf(SandboxService("mcp", 8080u, null)))
    val transport = ScriptedHttpClient()
    val credentials = CyclopsCredentials("client-id", "client-secret")
    val client = CyclopsClient.connect(CyclopsConfiguration("https://cyclops.invalid", "https://keycloak.invalid/token", credentials, 1uL, 1u, 1uL, 2u), transport)
    val pool = client.createPool(CreatePoolRequest("default", spec))
    val claim = client.createClaim(CreateClaimRequest(pool, ClaimSpec(SandboxTemplateRef(pool.metadata.name), null, null, null)))
    val sandbox = client.waitClaim(claim)
    val service = client.serviceRequest(sandbox, "mcp", "/mcp", HttpRequest("POST", "https://ignored.invalid/mcp", emptyList(), "{\"offline\":true}".encodeToByteArray()))
    client.deleteClaim(claim)
    client.deletePool(pool)
    check(pool.metadata.name == "default" && claim.metadata.name == "default" && sandbox.name == "offline-sandbox")
    check(service.status == 202.toUShort())
    transport.assertExhausted()
    println("pool=${pool.metadata.name} claim=${claim.metadata.name} sandbox=${sandbox.name} service_status=${service.status}")
}
